probe-rs/src/debug/mod.rs

```diff
@@ -723,94 +904,348 @@ impl DebugInfo {
+                        referenced_variable.parent_key,
+                        referenced_variable,
+                        core,
+                    )?;
+                    referenced_variable = unit_info.extract_type(
+                        referenced_node,
+                        &parent_variable,
+                        referenced_variable,
+                        core,
+                        stack_frame_registers,
+                    )?;
+
+                    // Only use this, if it is NOT a unit datatype.
+                    if referenced_variable.type_name.contains("()") {
+                        self.variable_cache
+                            .remove_cache_entry(referenced_variable.variable_key)?;
+                    } else if parent_variable.name.eq("<statics>") {
+                        // If we are lazily resolving `<statics>`, then we need to eliminate the intermediate node
+                        self.variable_cache
+                            .adopt_grand_children(&parent_variable, &referenced_variable)?;
+                    }
+                }
+            }
         }
+        Ok(())
     }
 
     /// Returns a populated (resolved) [`StackFrame`] struct.
+    /// This function will also populate the `DebugInfo::VariableCache` with in scope `Variable`s for each `StackFrame`
     fn get_stackframe_info(
         &self,
         core: &mut Core<'_>,
         address: u64,
-        frame_count: u64,
-        registers: Registers,
-        inlined_function: bool,
+        unwind_registers: &Registers,
     ) -> Result<StackFrame, DebugError> {
         let mut units = self.get_units();
 
         let unknown_function = format!("<unknown function @ {:#010x}>", address);
+        let mut stack_frame_registers = unwind_registers.clone();
 
         while let Some(unit_info) = self.get_next_unit_info(&mut units) {
-            if let Some(die_cursor_state) =
-                &mut unit_info.get_function_die(address, inlined_function)
-            {
-                let function_name = die_cursor_state
+            if let Some(function_die) = &mut unit_info.get_function_die(address, true) {
+                let function_name = function_die
                     .function_name(&unit_info)
                     .unwrap_or(unknown_function);
 
-                log::debug!("Function name: {}", function_name);
+                if function_die.is_inline {
+                    // Calculate the call site for this function, so that we can use it later to create an additional 'callee' `StackFrame` from that PC.
+                    let address_size =
+                        gimli::_UnwindSectionPrivate::address_size(&self.frame_section) as u64;
+                    if function_die.low_pc > address_size && function_die.low_pc < u32::MAX.into() {
+                        stack_frame_registers
+                            .set_return_address(Some((function_die.low_pc - address_size) as u32));
```

Comment:
I have a case where unwinding an inline function fails completely:

```
0: __wfe() inline
   ./asm/lib.rs:48

1: __udf() inline
   ./asm/lib.rs:48

2: __sev() inline
   ./asm/lib.rs:48

3: __primask_r() inline
   ./asm/inline.rs:156

4: __cpsie() inline
   ./asm/lib.rs:48

5: __cpsid() inline
   ./asm/lib.rs:48

6: try_from()
   /home/dominik/code/stm32wb55-embassy/src/tl_mbox/consts.rs

7: wfe() inline
   /home/dominik/.cargo/registry/src/github.com-1ecc6299db9ec823/cortex-m-0.7.4/src/asm.rs:49

```

On master, it looks like this:

```
537066440: __wfe (inline)
	./asm/lib.rs:48

537066440: __wfe
	./asm/lib.rs:51:17

537066440: poll
	/rustc/c5ecc157043ba413568b09292001a4a74b541a4e/library/core/src/sync/atomic.rs:2364:24

537066440: run<stm32wb55_embassy::__cortex_m_rt_main::{closure#0}>
	/home/dominik/.cargo/git/checkouts/embassy-f7d634104174c60a/1a6b0b5/embassy/src/executor/arch/cortex_m.rs:53:22

537066456: __cortex_m_rt_main
	/home/dominik/code/stm32wb55-embassy/src/main.rs:177:1

537066488: __cortex_m_rt_main_trampoline
	/home/dominik/code/stm32wb55-embassy/src/main.rs:177:1

6: <unknown function @ 0x08000180>
	/home/jonas/dev/cortex-m-rt/asm.S:106
```

For context, this is the DWARF for the original function:

```
0x0002be9c:       DW_TAG_subprogram
                            DW_AT_low_pc  (0x080034ae)
                            DW_AT_high_pc (0x080034b2)
                            DW_AT_frame_base  (DW_OP_reg7 R7)
                            DW_AT_name    ("__wfe")
                            DW_AT_decl_file   ("./asm/lib.rs")
                            DW_AT_decl_line   (48)
                            DW_AT_external    (true)
    
0x0002bead:         DW_TAG_inlined_subroutine
                            DW_AT_abstract_origin   (0x0002b9a5 "_ZN3lib6inline5__wfe17h869759b7613e9d29E")
                            DW_AT_low_pc    (0x080034ae)
                            DW_AT_high_pc   (0x080034b0)
                            DW_AT_call_file ("./asm/lib.rs")
                            DW_AT_call_line (51)
                            DW_AT_call_column   (0x11)
```
For me, the initial two frames look correct on the master branch, but on this PR something goes wrong, and it includes different unrelated frames in the stack trace.

Reply:
> @Tiwalun You are correct. This code generates a 'virtual' stack frame , to allow you to debug the inlined function, and the effect ass you pointe out, is that the virtual LR value is not 'real'.
> 
> Do you have suggestions for how we can make that obvious to the observer, or even do it differently?

I think we should just not touch the return address. For unwinding, we don't have to change it, so I don't see any advantage in doing this.

Reply:
If I manually ignore inlined functions, the stack trace from the PR looks like this:

```
0: __wfe()
   ./asm/lib.rs:48 x

1: run<stm32wb55_embassy::__cortex_m_rt_main::{closure#0}>()
   /home/dominik/.cargo/registry/src/github.com-1ecc6299db9ec823/cortex-m-0.7.4/src/asm.rs:49

2: __cortex_m_rt_main()
   /home/dominik/code/stm32wb55-embassy/src/main.rs:177

3: __cortex_m_rt_main_trampoline()
   /home/dominik/code/stm32wb55-embassy/src/main.rs:177
```

This looks a lot more plausible, so I guess something is broken with the handling of inlined functions.

Reply:
Thanks, this helps me to understand what you are running into. Unfortunately, due to some unforeseen circumstances, I do not have access to any dev boards to test against this week. I'd really like to reproduce and resolve this issue before we merge this PR. I will pick it up on Monday again and get back to you if I have more questions.

Reply:
@Tiwalun , I have cleaned up the incorrect setting of LR register for inlined functions, as well as improved how the `StackFrames` are created. I would appreciate if you could test this on your test case to validate my results.

Here is what I get ... in A) below, you can see the file and line/column numbers for the inlined function call, as well as the register values.  in B) below, I have exactly the same code, but the function IS NOT inlined during build (doing this so that you can see file/line/column matches up.

A) wfe() is inlined during build
<img width="713" alt="Screen Shot 2022-02-03 at 10 55 32 AM" src="https://user-images.githubusercontent.com/5413046/152382315-8ca8cacb-727a-42d9-9db0-7591cb7e8b87.png">

B) wfe() is not inlined during build 
<img width="721" alt="Screen Shot 2022-02-03 at 10 57 50 AM" src="https://user-images.githubusercontent.com/5413046/152382457-cce4cd33-5d15-4275-82a5-ab7c42f7ef08.png">

Reply:
Seems fine now. :+1: